### PR TITLE
Saving and loading snapshots without temp files

### PR DESCRIPTION
**Pull request by monkeyman79**
- Added functions `int machine_write_snapshot_to_stream(snapshot_stream_t *stream, int save_roms,
int save_disks, int event_mode)` and `int machine_read_snapshot_from_stream(snapshot_stream_t *stream, int event_mode)` to each model
- Functions `snapshot_stream_t* snapshot_file_read_fopen(const char* pathname)` and `snapshot_stream_t* snapshot_file_write_fopen(const char* pathname)` creates `snapshot_stream_t` object using regular `FILE`
- Functions `snapshot_stream_t* snapshot_memory_read_fopen(const void* buffer, size_t buffer_size)` and `snapshot_stream_t* snapshot_memory_write_fopen(void* buffer, size_t buffer_size)` create memory based stream
- Call `snapshot_memory_write_fopen(NULL, 0)` creates dummy memory stream which is used for calculating snapshot size

### CODE REVIEW

**Comment by rsn8887** [comment]
Awesome, let me include this in my tests on Switch also. I will report in the next few days, and merge if there are no problems with it.

**Comment by rsn8887** [comment]
Can you solve the conflicts? They arose from merging your first PR. Thanks! I am eager to test this, because this would a great improvement!

**Comment by monkeyman79** [comment]
Is this OK, or should I (try to) squash it?

**Comment by monkeyman79** [comment]
Bad - let me try again

**Comment by monkeyman79** [comment]
That's better

**Comment by rsn8887** [comment]
It works great! Thanks, merging.

**Comment by snoofly** [comment]
Nice. Great work all.

**Comment by snoofly** [comment]
Nice. Great work all.
